technology_and_tooling/testing/mocking.md

```diff
@@ -0,0 +1,526 @@
+    """Mock the database connection and cursor to ensure the correct methods
+    are called within the query_database function
+    """
+    with patch('sqlite3.connect') as mock_connection:
+        mock_cursor = mock_connection.return_value.cursor.return_value
+        mock_cursor.fetchall.return_value = [("Jerry", "Mouse", 1)]
+        # make our fake connection
+        conn = mock_connection("my_non_existent_file")
+        sql = "SELECT * FROM Animals"
+        # test what query_database does with our fake connection
+        result = query_database(sql, connection=conn)
+        assert result[0] == ("Jerry", "Mouse", 1)
+        # check that query_database passes our SQL string to cursor.execute()
+        mock_cursor.execute.assert_called_once_with(sql)
+        # check that fetchall() was called
+        mock_cursor.fetchall.assert_called_once()
+        # check that query_database closes the connection
+        conn.close.assert_called_once()
+~~~
+
+:::
+::::
+
+Now lets look at using a function decorator.
+
+::::challenge{id=patch-decorator title="Using the `@patch` decorator"}
+
+Rewrite `test_query_db_mocked_connection` to use a function decorator instead of a context manager. You can view the [unittest documentation](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) as a guide.
+
+:::solution
+~~~python
+@patch('sqlite3.connect')
+def test_query_db_mocked_connection(mock_connection):
+    mock_cursor = mock_connection.return_value.cursor.return_value
+    mock_cursor.fetchall.return_value = [("Jerry", "Mouse", 1)]
+    # make our fake connection
+    conn = mock_connection("my_non_existent_file")
+    sql = "SELECT * FROM Animals"
+    # test what query_database does with our fake connection
+    result = query_database(sql, connection=conn)
+    assert result[0] == ("Jerry", "Mouse", 1)
+    # check that query_database passes our SQL string to cursor.execute()
+    mock_cursor.execute.assert_called_once_with(sql)
+    # check that fetchall() was called
+    mock_cursor.fetchall.assert_called_once()
+    # check that query_database closes the connection
+    conn.close.assert_called_once()
+~~~
+
+:::
+::::
+
+### Mocking using other libraries
+
+#### Monkeypatching in pytest
+
+As an alternative to using the `unitest.mock` library, its possible to use a version of mocking from within `pytest`, termed *monkeypatching*. A built-in fixture called `monkeypatch` allows modifying attributes, functions or classes within the scope of the test function. Some example methods are:
+
+- `monkeypatch.setattr()` - used to set an attribute to a new value or replace it with a new function
+- `monkeypatch.delattr()` - used to delete an attribute
+- `monkeypatch.setitem()` - used to modify a dictionary item
+- `monkeypatch.delitem()` - used to delete an item from a dictionary
+
+Here is an example of how we could replace `unitest.mock` with `monkeypatch` in our `test_query_db_mocked_connection()` test:
```

Comment:
again, it seems like `monkeypatch` is a replacement for `unittest.mock.patch`, not the whole `unittest.mock` library. Below you show how to use them both in combination, I would just start using them in combination from the start if that is how they are designed to be used.

Reply:
I've changed the text to just use them together now, thanks